### PR TITLE
Copilot Chat - Bot portability

### PR DESCRIPTION
**Pull request by Vivihung**
### Motivation and Context
This change enables download/upload a bot in the Copilot Chat app. It is the first step to demonstrate the idea of "bots are documents". With Semantic Kernel memory and chat history, users can authorize/update/share bots with others, like write a document. We would like to inspire SK developers with this example and learn from their real-world use cases in order to take the idea further.

### Description
* UX controls to trigger download/upload a bot file. Now we only support download/upload files from the user's computer.
* API endpoints to expose the upload/download functionality.
    * Download: Export 1) chat message history from the given chat session (id: chat id), 2) embeddings from SK memory store.
    * Upload: Import 1) chat message history to a new chat session, 3) embeddings to SK memory store. (Note: Copilot Chat messages are not linked to SK memory id.)
* Bot schema v1.0. It defines what is stored in a bot file - 1) schema version info (for compatibility), 2) chat history, 3) embeddings from the conversation, 4) the AI service information that generated the embeddings.
* Error handlings:
    1. If schema in the file and in the app config mismatch, the API throws exception. For embedding mismatches, further enhancement could be done to generate new embeddings.
    2. If import failed in any step, we throw an exception. A further work item here to revert changes in each store to make it transactional.
* Chat history: Hallucination is possible if the model missed context. Therefore, we store and render the chat history in the experience to allow users to tell if hallucination is presented for a given prompt.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [] ~All unit tests pass, and I have added new tests where possible~
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: samples/apps/copilot-chat-app/webapi/Config/BotSchemaConfig.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    /// <summary>
+    /// The name of the schema.
+    /// </summary>
+    public string Name { get; set; } = string.Empty;
+
+    /// <summary>
+    /// The version of the schema.
+    /// </summary>
+    public int Verson { get; set; } = -1;
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Config/BotSchemaConfig.cs:18]:
```suggestion
    public int Version { get; set; } = -1;
```

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                await this._chatMessageRepository.CreateAsync(chatMessage);
+            }
+
+            // 3. Update the memory.
+            await this.BulkUpsertMemoryRecordsAsync(oldChatId, chatId, bot.Embeddings);
+        }
+        catch
+        {
+            // TODO: Revert changes if any of the actions failed
+            throw;
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:94]: If we're not going to handle the errors, let's just have the TODO inline and remove the try/catch.

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        }
+
+        return this.Accepted();
+    }
+
+    /// <summary>
+    /// Download a bot.
+    /// </summary>
+    /// <param name="kernel"></param>
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:103]: XMLDoc for parameter descriptions

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+using Microsoft.SemanticKernel.Memory;
+using SemanticKernel.Service.Config;
+using SemanticKernel.Service.Model;
+using SemanticKernel.Service.Skills;
+using SemanticKernel.Service.Storage;
+
+namespace SemanticKernel.Service.Controllers;
+
+[ApiController]
+public class BotController : ControllerBase
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:16]: We call these "Conversations" in the UX. Could we stay consistent with that terminology throughout?

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        [FromServices] Kernel kernel,
+        [FromQuery] string userId,
+        [FromBody] Bot bot)
+    {
+        // TODO: We should get userId from server context instead of from request for privacy/security reasons when support multipe users.
+        this._logger.LogDebug("Received call to upload a bot");
+
+        if (!this.IsBotCompatible(bot.Schema, bot.EmbeddingConfigurations))
+        {
+            return new UnsupportedMediaTypeResult();
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:64]: It is difficult for the caller to understand why it received this result, which is normally reserved for handling mismatched application content types (e.g. json vs image). Consider returning BadRequest with a content message explaining that the embedding configurations are incompatible.

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    [Route("bot/upload")]
+    [ProducesResponseType(StatusCodes.Status202Accepted)]
+    [ProducesResponseType(StatusCodes.Status415UnsupportedMediaType)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult> UploadAsync(
+        [FromServices] Kernel kernel,
+        [FromQuery] string userId,
+        [FromBody] Bot bot)
+    {
+        // TODO: We should get userId from server context instead of from request for privacy/security reasons when support multipe users.
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:59]: Is this relevant for this API call?

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    /// <summary>
+    /// Check if an external bot file is compatible with the application.
+    /// </summary>
+    /// <remarks>
+    /// If the embeddings are not generated from the same model, the bot file is not compatible.
+    /// </remarks>
+    /// <param name="externalBotSchema">The external bot schemal.</param>
+    /// <param name="externalBotEmbeddingConfig">The external bot embedding configuration.</param>
+    /// <returns></returns>
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:130]: [nit] Populate or remove empty XMLDocs

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    private bool IsBotCompatible(BotSchemaConfig externalBotSchema, BotEmbeddingConfig externalBotEmbeddingConfig)
+    {
+        var embeddingAIServiceConfig = this._configuration.GetSection("Embedding").Get<AIServiceConfig>();
+        var botSchema = this._configuration.GetSection("BotFileSchema").Get<BotSchemaConfig>();
+
+        if (embeddingAIServiceConfig != null && botSchema != null)
+        {
+            // The app can define what schema/version it supports before the community comes out with an open schema.
+            return externalBotSchema.Name.Equals(
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:139]: this indenting is obfuscating the logic here. 
```
return externalBotSchema.Name.Equals(botSchema.Name, StringComparison.OrdinalIgnoreCase) &&
    externalBotSchema.Verson == botSchema.Verson &&
    externalBotEmbeddingConfig.AIService.Equals(embeddingAIServiceConfig.AIService, StringComparison.OrdinalIgnoreCase) &&
    externalBotEmbeddingConfig.DeploymentOrModelId.Equals(embeddingAIServiceConfig.DeploymentOrModelId, StringComparison.OrdinalIgnoreCase);
```

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    {
+        // TODO: We might want to set limitation on the number of messages that are pulled from the storage.
+        return (await this._chatMessageRepository.FindByChatIdAsync(chatId))
+            .OrderByDescending(m => m.Timestamp).ToList();
+    }
+
+    /// <summary>
+    /// Bulk upsert memory records into memory store.
+    /// </summary>
+    /// <param name="oldChatId"></param>
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:224]: XMLDoc

File: samples/apps/copilot-chat-app/webapi/Model/Bot.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+using Microsoft.SemanticKernel.Memory;
+using SemanticKernel.Service.Config;
+using SemanticKernel.Service.Skills;
+
+namespace SemanticKernel.Service.Model;
+
+/// <summary>
+/// The data model of a bot for portability.
+/// </summary>
+public class Bot
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Model/Bot.cs:12]: Recommend "Conversation" or "Session".

File: samples/apps/copilot-chat-app/webapi/Skills/ChatMessage.cs [hunk 11]
```diff
@@ -19,6 +19,7 @@ public enum AuthorRoles
     {
         User = 0,
         Bot,
+        Participant
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapi/Skills/ChatMessage.cs:22]: What is the participant in this context? Maybe its a good time to add a comment describing each role.

File: samples/apps/copilot-chat-app/webapp/src/libs/models/ChatMessage.ts [hunk 12]
```diff
@@ -3,6 +3,7 @@
 export enum AuthorRoles {
     User = 0,
     Bot,
+    Participant
```
Comment (adrianwyatt) [inline samples/apps/copilot-chat-app/webapp/src/libs/models/ChatMessage.ts:6]: Could we add a comment describing each role?

**Review by adrianwyatt — CHANGES_REQUESTED** [review]

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 2 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:94]: I left the TODO there to handle the error in a separate PR.

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 4 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:16]: I understand your points. I wonder how that aligns with the concept of "bots are docs". hm... "conversations are docs" doesn't communicate the disruptive idea, IMO...

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 8 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:139]: Good catch! I don't recall what caused the weird indent. But I do agree with you on readability.

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 5 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs]: Very good feedback. thanks!

**Review by adrianwyatt — APPROVED** [review]

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 7 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:130]: good catch. Do you have a tool to spot missing XMLDocs? If so, I would love to setup one in my dev env.

File: samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs (hunk 6 above)
Comment (Vivihung) [reply samples/apps/copilot-chat-app/webapi/Controllers/BotController.cs:59]: Yes, it's for the userId that we currently get from the request query string. I've log a task to address those as a refactory.

**Pull request merged**